device-api/src/list.rs

```diff
@@ -3,62 +3,68 @@ use std::fs::FileType;
 use std::io;
 use std::path::{Path, PathBuf};
 
+use strum::IntoEnumIterator;
 use tokio::fs;
 
+use crate::arch::Arch;
 use crate::devfs::{self, is_character_device};
-use crate::device::{Device, DeviceFile, DeviceInfo};
+use crate::device::{Device, DeviceFile, DeviceInner};
 use crate::error::DeviceResult;
-use crate::sysfs::npu_mgmt::{self, *};
+use crate::sysfs::npu_mgmt;
 use crate::{hwmon, DeviceError};
 
 /// Allow to specify arbitrary sysfs, devfs paths for unit testing
 pub(crate) async fn list_devices_with(devfs: &str, sysfs: &str) -> DeviceResult<Vec<Device>> {
-    let npu_dev_files = filter_dev_files(list_devfs(devfs).await?)?;
-
-    let mut devices: Vec<Device> = Vec::with_capacity(npu_dev_files.keys().len());
-
-    for (idx, paths) in npu_dev_files {
-        if let Ok(device) = get_device_inner(idx, paths, devfs, sysfs).await {
-            devices.push(device);
+    let mut devices = Vec::new();
+    let mut idx: u8 = 0;
+    for arch in Arch::iter() {
+        let dev_files = match list_dev_files(arch.devfile_path(devfs)).await {
+            Ok(files) => files,
+            Err(e) if e.kind() == io::ErrorKind::NotFound => Vec::new(),
+            Err(e) => return Err(e.into()),
+        };
+        let mut npu_dev_files_sorted = filter_dev_files(dev_files)?.into_iter().collect::<Vec<_>>();
+        npu_dev_files_sorted.sort_by_key(|(idx, _)| *idx);
+
+        for (_, paths) in npu_dev_files_sorted {
+            if let Ok(device) = get_device_inner(arch, idx, paths, devfs, sysfs).await {
+                devices.push(device);
+            }
+            idx += 1;
```

Comment:
I don't know why you are using the `idx` variable in the outer for loop, instead of using the `idx` contained in `npu_dev_files_sorted`. Can you explain your intent?

Reply:
There was a long conversation between me and @bg-furiosa, and the code above follows the requirements we agreed. Please check out https://github.com/furiosa-ai/device-api/pull/106#discussion_r1560419897, and any additional feedback is welcome.

Reply:
As per the offline discussion we had yesterday, I reverted this change and renamed device_index to devfile_index.

[abc3937](https://github.com/furiosa-ai/device-api/pull/106/commits/abc3937555bc3ba43f6e40a8e043755b78289ac4)
[e171f6b](https://github.com/furiosa-ai/device-api/pull/106/commits/e171f6bf18ca0e60c1ff77cc0e73afdc6de8a6ed)

Reply:
My bad, there still remains some code that depends on `device_index`, will fix it now.

Reply:
[e3aa449](https://github.com/furiosa-ai/device-api/pull/106/commits/e3aa44933a4397a6c3080c2fbefdd04422ae72aa)

Also changed all the remaining code that tried to display the device as `npu{idx}`.